libkbfs/tlf_journal.go

```diff
@@ -1238,11 +1360,90 @@ func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
 		return MdID{}, err
 	}
 
+	// Merge the unflushed paths.
+	if newUnflushedPaths != nil {
+		if _, ok := j.unflushedPaths[rmd.Revision()]; ok {
+			return MdID{}, fmt.Errorf("Revision %d already has unflushed paths",
+				rmd.Revision())
+		}
+		j.unflushedPaths[rmd.Revision()] = newUnflushedPaths
+	}
+
 	j.signalWork()
 
 	return mdID, nil
 }
 
+// prepUnflushedPaths returns a set of paths that were updated in this
+// revision.  The caller must NOT already hold `journalLock`.
+func (j *tlfJournal) prepUnflushedPaths(ctx context.Context,
+	rmd *RootMetadata) (map[string]bool, error) {
+	cpp := func() chainsPathPopulator {
+		j.journalLock.RLock()
+		defer j.journalLock.RUnlock()
+		return j.chainsPopulator
+	}()
+
+	// The unflushed paths haven't been initialized yet.
+	if cpp == nil {
+		return nil, nil
+	}
+
+	newUnflushedPaths := make(map[MetadataRevision]map[string]bool)
+	irmds := []ImmutableRootMetadata{
+		// Revision and timestamp don't matter for finding the
+		// unflushed paths.
+		MakeImmutableRootMetadata(rmd, fakeMdID(1), time.Now()),
+	}
+
+	err := j.addUnflushedPaths(ctx, irmds, cpp, newUnflushedPaths)
+	if err != nil {
+		return nil, err
+	}
+	if len(newUnflushedPaths) > 1 {
+		return nil, fmt.Errorf("%d unflushed revisions on a single put",
+			len(newUnflushedPaths))
+	}
+
+	return newUnflushedPaths[rmd.Revision()], nil
+}
+
+func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
+	MdID, error) {
+	// Before we take the lock, get the unflushed paths ready (which
+	// might require taking the lock).
+	newUnflushedPaths, err := j.prepUnflushedPaths(ctx, rmd)
+	if err != nil {
+		return MdID{}, err
+	}
+	mdID, err := j.doPutMD(ctx, rmd, newUnflushedPaths)
+
+	// If someone initialized j.unflushedPaths, then try prepping the
+	// unflushed paths one more time.
+	if err == errRetryUnflushedPaths {
+		err := j.maybeWaitForUnflushedReady(ctx)
```

Comment:
(Oops, hit "Comment' too early above.)

Ah, I see now you mentioned a `removeFromCache` method.

Otherwise, I don't see any obvious concurrency differences with your model,so I'll give it a try.  Thanks for the suggestions.